README.md
<div><h1>pleasure-di</h1></div>

<p>
    <a href="https://www.npmjs.com/package/pleasure-di" target="_blank"><img src="https://img.shields.io/npm/v/pleasure-di.svg" alt="Version"></a>
<a href="http://opensource.org/licenses" target="_blank"><img src="http://img.shields.io/badge/License-MIT-brightgreen.svg"></a>
</p>

<p>
    a simple dependency injection module
</p>

## Installation

```sh
$ npm i pleasure-di --save
# or
$ yarn add pleasure-di
```

## Features

- [register containers](#register-containers)
- [injects dependencies](#injects-dependencies)
- [provides an interface to customize the container resolution](#provides-an-interface-to-customize-the-container-resolution)


<a name="register-containers"></a>

## register containers


```js
const container = pleasureDi({
  Gateway (name) {
    if (name === 'NotFoundGateway') {
      return
    }

    return () => {}
  }
})

t.throws(() => container.SomeService, {
  message: 'No suffixed container name registered matching SomeService'
})

t.throws(() => container.NotFoundGateway, {
  message: 'Resource NotFoundGateway not found'
})
t.notThrows(() => container.SomeGateway)
```

<a name="injects-dependencies"></a>

## injects dependencies


```js
const container = pleasureDi({
  Gateway (gatewayName) {
    // holds the logic that resolves given gatewayName
    // returns a function that will come with the container
    return (/* container */) => {
      return {
        methodA (caller) {
          return `methodA from ${gatewayName} called by ${caller}`
        },
        methodB (caller) {
          return `methodB from ${gatewayName} called by ${caller}`
        }
      }
    }
  },
  Service (serviceName) {
    // holds the logic that resolves given serviceName
    // returns a function that will come with the container
    return ({ FreakingGateway }) => {
      return {
        methodC () {
          return FreakingGateway.methodA(serviceName)
        },
        methodD () {
          return FreakingGateway.methodB(serviceName)
        }
      }
    }
  }
})

const { PaymentService } = container

t.is(PaymentService.methodC(), 'methodA from FreakingGateway called by PaymentService')
t.is(PaymentService.methodD(), 'methodB from FreakingGateway called by PaymentService')
```

<a name="provides-an-interface-to-customize-the-container-resolution"></a>

## provides an interface to customize the container resolution


```js
const resolveFixture = (...fixturePath) => {
  return path.join(__dirname, '__tests__/fixtures', ...fixturePath)
}
const container = pleasureDi({
  Gateway (gatewayName) {
    return require(resolveFixture('gateways', gatewayName) + '.js')
  },
  Repository (respositoryName) {
    return require(resolveFixture('repositories', respositoryName) + '.js')
  },
  Service (serviceName) {
    return require(resolveFixture('services', serviceName) + '.js')
  }
})

const { OrderService } = container

t.deepEqual(OrderService.findByUserId(123), {
  id: 123
})
t.is(OrderService.pay(1111), 'amount 1111 paid')
t.throws(() => container.SomeService, {
  code: 'MODULE_NOT_FOUND'
})
```


<br><a name="pleasureDi"></a>

### pleasureDi(resolvers) ⇒ <code>Object</code> \| <code>\*</code>

| Param | Type | Description |
| --- | --- | --- |
| resolvers | <code>Object</code> | resolvers map |


<br><a name="Resolved"></a>

### Resolved ⇒ <code>Object</code>

| Param | Type |
| --- | --- |
| container | <code>Object</code> | 


<br><a name="Resolver"></a>

### Resolver ⇒ [<code>Resolved</code>](#Resolved)

| Param | Type |
| --- | --- |
| containerName | <code>String</code> | 


* * *

### License

[MIT](https://opensource.org/licenses/MIT)

&copy; 2020-present Martin Rafael Gonzalez <tin@devtin.io>


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.1.0](https://github.com/devtin/pleasure-di/compare/v1.0.0...v1.1.0) (2021-02-09)


### Features

* avoid promises ([bc26caf](https://github.com/devtin/pleasure-di/commit/bc26cafa454d50342afd3772ea6fbb6d860cc26e))

## 1.0.0 (2021-02-06)


src/lib/__tests__/fixtures/services/OrderService.js
module.exports = ({ PayPalGateway, OrderRepository }) => {
  return {
    pay (amount) {
      return PayPalGateway.pay(amount)
    },
    findByUserId (userId) {
      return OrderRepository.findByUserId(userId)
    }
  }
}


src/lib/__tests__/fixtures/gateways/PayPalGateway.js
module.exports = () => {
  return {
    pay (amount) {
      return `amount ${amount} paid`
    }
  }
}


src/lib/__tests__/fixtures/repositories/OrderRepository.js
module.exports = () => {
  return {
    findByUserId (id) {
      return {
        id
      }
    }
  }
}


src/lib/pleasure-di.js
const each = (arr, cb) => {
  for (let i = 0; i < arr.length; i++) {
    if (cb(arr[i], i) === false) {
      break
    }
  }
}

/**
 * @typedef {Function} Resolved
 * @param {Object} container
 * @return {Object}
 */

/**
 * @typedef {Function} Resolver
 * @param {String} containerName
 * @return {Resolved}
 */

/**
 *
 * @param {Object} resolvers - resolvers map
 * @return {{}|*}
 */

export const pleasureDi = (resolvers) => {
  const availableContainers = Object.keys(resolvers)
  const containerMayExist = (containerNameRequested) => {
    let mayExist = false
    each(availableContainers, (containerNameRegistered) => {
      const containerNameMatcher = new RegExp(`${containerNameRegistered}$`)
      if (containerNameMatcher.test(containerNameRequested)) {
        mayExist = containerNameRegistered
        return false
      }
    })
    return mayExist
  }


  const container = new Proxy({}, {
    get (target, path) {
      // avoid promises
      if (path === 'then') {
        return
      }

      const matchingContainer = containerMayExist(path)
      if (!matchingContainer) {
        throw new Error(`No suffixed container name registered matching ${path}`)
      }

      const resolver = resolvers[matchingContainer](path)

      if (!resolver) {
        throw new Error(`Resource ${path} not found`)
      }

      return resolver(container)
    }
  })

  return container
}


src/lib/pleasure-di.unit.js
import test from 'ava'
import path from 'path'
import { pleasureDi } from './pleasure-di.js'

test('register containers', (t) => {
  const container = pleasureDi({
    Gateway (name) {
      if (name === 'NotFoundGateway') {
        return
      }

      return () => {}
    }
  })

  t.throws(() => container.SomeService, {
    message: 'No suffixed container name registered matching SomeService'
  })

  t.throws(() => container.NotFoundGateway, {
    message: 'Resource NotFoundGateway not found'
  })
  t.notThrows(() => container.SomeGateway)
})

test('injects dependencies', (t) => {
  const container = pleasureDi({
    Gateway (gatewayName) {
      // holds the logic that resolves given gatewayName
      // returns a function that will come with the container
      return (/* container */) => {
        return {
          methodA (caller) {
            return `methodA from ${gatewayName} called by ${caller}`
          },
          methodB (caller) {
            return `methodB from ${gatewayName} called by ${caller}`
          }
        }
      }
    },
    Service (serviceName) {
      // holds the logic that resolves given serviceName
      // returns a function that will come with the container
      return ({ FreakingGateway }) => {
        return {
          methodC () {
            return FreakingGateway.methodA(serviceName)
          },
          methodD () {
            return FreakingGateway.methodB(serviceName)
          }
        }
      }
    }
  })

  const { PaymentService } = container

  t.is(PaymentService.methodC(), 'methodA from FreakingGateway called by PaymentService')
  t.is(PaymentService.methodD(), 'methodB from FreakingGateway called by PaymentService')
})

test('provides an interface to customize the container resolution', (t) => {
  const resolveFixture = (...fixturePath) => {
    return path.join(__dirname, '__tests__/fixtures', ...fixturePath)
  }
  const container = pleasureDi({
    Gateway (gatewayName) {
      return require(resolveFixture('gateways', gatewayName) + '.js')
    },
    Repository (respositoryName) {
      return require(resolveFixture('repositories', respositoryName) + '.js')
    },
    Service (serviceName) {
      return require(resolveFixture('services', serviceName) + '.js')
    }
  })

  const { OrderService } = container

  t.deepEqual(OrderService.findByUserId(123), {
    id: 123
  })
  t.is(OrderService.pay(1111), 'amount 1111 paid')
  t.throws(() => container.SomeService, {
    code: 'MODULE_NOT_FOUND'
  })
})
